### DISCUSSION TITLE
Best practices for optimizing usage of simdjson

### DISCUSSION THREAD

**Discussion by spnda** [simdjson/simdjson#1983] [General]
While writing fastgltf (thanks for putting it into the README!) I have become concerned that the way I use the API and structure my parsing code affects the performance of the library. I do still get better performance than other glTF libraries that use RapidJSON or [jsmn](https://github.com/zserge/jsmn), though have noticed that it is nowhere near the raw parsing benchmarks found in the README and doc/. I would suspect that perhaps the way I use the API might be impacting the real world usage.

I am aware of `doc/performance.md` but do think that it should also mention if the error handling method chosen affects performance in any way or if one should watch out for anything in the API.

I for one currently don't use exceptions, though nowhere can I find any specific benchmarks that compare the two methods of checking for errors if either one is quicker. I also essentially always get a value with `object["key"].get_string().get(str)` and then directly create a new `std::string` with that view (which allocates) to some struct if that returns SUCCESS, and then skip to the next one. Is there perhaps some problem with repeatedly switching between SIMD instructions and normal x86 instructions that could cause any significant slowdown?

As I knew that using unordered find field with ondemand would cause the iterator to possibly parse the object's json up to N times, where N is the amount of fields I'm searching for, and the fact that with the glTF format there is no guarantee that the fields have any order (seems like every exporter just does it their own way) so I decided that I'd just loop over all fields once, hash the key I encounter and then use a fat switch to parse the correct object. Is that recommended? Is there any better way with the API to handle that case?

Furthermore, I have noticed that using ondemand is actually slower than dom in my case. I've encountered cases where ondemand is 28% slower, while in other cases it is just 3% slower, though it is never quicker. This might be because I just took the same code and replaced dom with ondemand, changing virtually nothing else. I can't find anything on if there are different design choices one should use between the two approaches. This kinda doesn't make sense to me because I'd think dom means it has to allocate some amount of memory to store all the information so I'd automatically figure that it should be slower, but it just isn't.

Seriously don't know if it matters, but I am using a Zen 3 chip (5800X). I've had the thought that perhaps AMD's SIMD instructions have a higher overhead than on Intel and therefore only give a small margin over other parsers, but that thought has no facts to back it up.

fastgltf is also openly available so you could go in and read [`src/fastgltf.cpp`](https://github.com/spnda/fastgltf/blob/main/src/fastgltf.cpp). Perhaps you notice something that I'm doing really wrong that is causing any sort of slowdown. I'd greatly appreciate any help!

Though I'd generally love to see some documentation about best practices that allow people to see what not to do vs what to do, and some general notes and warnings about some things in the API.

**Comment by lemire** [comment]
@spnda It is fine to use the DOM and we have good reasons to support both On Demand and DOM. If you are loading everything and need, at least as an intermediate step, a database, DOM works well. It is also easier to use, somewhat, than On Demand.

You are obviously a good programmer. The code is quite decent. Still, there is a lot of code and it would be useful to have some profiling. Have you done any? E.g., perf record/perf report.

I am looking at https://github.com/spnda/fastgltf/pull/5/files

- I would not hardcode the assumption that SIMDJSON_BUILTIN_IMPLEMENTATION is always fallback. It is almost surely not hurting you but it is seems unnecessary and may indeed be wrong on some platforms.
- I would not use lines such as `animation.samplers.reserve(samplers.count_elements());` or `if (emissiveFactor.count_elements() != 3) {` unless you have determined that it is makes things faster. This requires the parser to travers the data once, rewind and come back again. Avoid any kind of `count_elements()` if you can. With the DOM API, this is likely making things faster, but maybe not with the DOM API.

Basically, the last point is my main concern. Try not depending so much on `count_elements()`. Also, try reporting back. I am quite interested in helping you out. But you know your code better than anyone. Can you try to identify the sore points to help us?

**Comment by lemire** [comment]
One nice trick also might be to run the code with logging enabled... (credit to @drcdr)

E.g., try setting `SIMDJSON_VERBOSE_LOGGING`. Then you can see in the log what On Demand does.

https://github.com/simdjson/simdjson/discussions/1938#discussioncomment-4743274


cc @jkeiser

**Comment by lemire** [comment]
I am currently adding some extra documentation (performance tips).

**Comment by lemire** [comment]
Note that simdjson is on godbolt which allows you to explore the assembly output:

https://godbolt.org/z/xE4GWs573

**Comment by lemire** [comment]
I added a performance tips section to our documentation: https://github.com/simdjson/simdjson/blob/master/doc/basics.md#performance-tips
